visualization/app/codeCharta/util/deltaGenerator.ts

```diff
@@ -100,23 +104,29 @@ export class DeltaGenerator {
 		}
 	}
 
-	private static getDeltaAttributeList(referenceAttribute: KeyValuePair, comparisonAttribute: KeyValuePair) {
-		const deltaAttribute: KeyValuePair = {}
+	private static compareAttributeValues(reference: KeyValuePair, comparison: KeyValuePair) {
+		const deltaList: KeyValuePair = {}
+		let differenceExists = false
 
-		// TODO: All entries should have the combined attributes and deltas set,
-		// even if they do not exist on one side. Calculate these attributes up
-		// front. This operation is otherwise costly.
-		for (const key of Object.keys(comparisonAttribute)) {
-			deltaAttribute[key] = referenceAttribute[key] ? comparisonAttribute[key] - referenceAttribute[key] : comparisonAttribute[key]
-		}
+		const attributeKeys = [...Object.keys(reference), ...Object.keys(comparison)]
```

Comment:
This might not be too bad because V8 has some nice optimizations under the hood for spreading arrays. In general, it's not ideal though, due to creating a new array that needs new memory. If we want to do this, we might also just use a set. That way there are no duplicate keys anymore and this improves the for loop below.

```suggestion
		const attributeKeys = new Set(Object.keys(reference))
		Object.keys(comparison).forEach((key) => attributeKeys.add(key))
```

Reply:
Yeah, that is a good point. I had a similar idea on Saturday morning, because the duplication seemed a bit itchy.

Reply:
No, a set works with hashes. Thus, each entry just calculates the hash and checks if it already exists. Calculating the hash is not for free but it is constant time.

Reply:
Resolved with commit [71bc9c6](https://github.com/MaibornWolff/codecharta/pull/2890/commits/71bc9c6a45bf3c33d7929b17e885c14646ae9888)